asf_core_data/pipeline/mcs/process/process_historical_mcs_installers.py

```diff
@@ -0,0 +1,926 @@
+    same_postcode_no_match = installer_data.groupby("postcode", as_index=False)[
+        ["company_unique_id"]
+    ].nunique()
+    same_postcode_no_match = same_postcode_no_match[
+        same_postcode_no_match["company_unique_id"] > 1
+    ]["postcode"].unique()
+
+    # New unique ID matches based on companies with same location
+    new_matches = map_installers_same_location_different_id(
+        same_postcode_no_match, installer_data
+    )
+    installer_data["company_unique_id"] = installer_data["company_unique_id"].apply(
+        lambda x: new_matches[x] if x in new_matches.keys() else x
+    )
+
+
+# copied from process_mcs_utils.py and changed slightly
+def geocode_postcode(data: pd.DataFrame, geodata: pd.DataFrame) -> pd.DataFrame:
+    """
+    Updates data with latitude and longitude columns, by merging with geodata
+    on postode column (left merge, not to loose any data).
+    Also transforms postcode column by removing the space.
+
+    Args:
+        data: DataFrame with postcode column.
+        geodata: DataFrame with postcode, latitude and longitude columns.
+    """
+
+    geodata["postcode"] = geodata["postcode"].str.replace(" ", "")
+
+    data["postcode"] = data["postcode"].str.upper().str.replace(" ", "")
+    data = data.merge(
+        geodata[["postcode", "latitude", "longitude"]], how="left", on="postcode"
+    )
+
+    return data
+
+
+def add_certification_body_info(
+    installer_data: pd.DataFrame, installations_data: pd.DataFrame
+) -> pd.DataFrame:
+    """
+    Adds certification body information to each instance of the historical installers table.
+
+    Historical installers data does not come with certification body information, only with certification number.
+    Company name and the certification number are used to merge and match to the installations table and get the
+    certification body from there.
+
+    Args:
+        installer_data: historical installers data
+        installations_data: historical installations data
+    Returns:
+        Updated table of installers
+    """
+
+    installations_match_vars = [
+        "installation_company_name",
+        "installation_company_mcs_number",
+    ]
+    installers_match_vars = ["company_name", "mcs_certificate_number"]
+
+    # dataframe with unique pairs (installation_company_name, installation_company_mcs_number)
+    # and corresponding certification body
+    unique_pairs_in_installations = installations_data.drop_duplicates(
+        installations_match_vars
+    )[
+        [
+            "installation_company_name",
+            "installation_company_mcs_number",
+            "certification_body",
+        ]
+    ]
+
+    # Merge installer with installation data to add the certification_body to the installer data
+    installer_data = installer_data.merge(
+        unique_pairs_in_installations,
+        how="left",
+        right_on=installations_match_vars,
+        left_on=installers_match_vars,
+    )
+
```

Comment:
When this function is called, the input installers and installations dataframes already have certification_body. Merge results in certification_body with _x and _y suffixes. Suggestion to assign _left and _right suffixes respectively for clarity, then merge _right data into _left if there are nulls in _left, using combine_first.

```
    # Merge installer with installation data to add the certification_body to the installer data
    installer_data = installer_data.merge(
        unique_pairs_in_installations,
        how="left",
        left_on=installers_match_vars,
        right_on=installations_match_vars,
        suffixes=('_left','_right')
    )

    # Unify _left and _right certification_body, filling any left nulls with right values
    installer_data["certification_body"] = \
        installer_data["certification_body_left"].combine_first(
            installer_data["certification_body_right"]
        )

Reply:
upsi! So sorry about this @sqr00t, I must have somehow (?) changed the location of the function call before adding the last commit. How weird, because all was running fine when I last run it.

I've changed the function call back to where it's supposed to be so that this doesn't happen. 

In short:
- The installers table doesn't originally have the certification body, that's why we do the merge to get the certification body from the installations table;
- However, when we get the missing installers data (from the installations table), we also get the certification body. So the getting the certification body for the original records should be done before we get the missing installers data.